// expo-cli/build/commands/eject.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function () {
    return data;
  };

  return data;
}

function _xdl() {
  const data = require("@expo/xdl");

  _xdl = function () {
    return data;
  };

  return data;
}

function _config() {
  const data = require("@expo/config");

  _config = function () {
    return data;
  };

  return data;
}

function Eject() {
  const data = _interopRequireWildcard(require("./eject/Eject"));

  Eject = function () {
    return data;
  };

  return data;
}

function LegacyEject() {
  const data = _interopRequireWildcard(require("./eject/LegacyEject"));

  LegacyEject = function () {
    return data;
  };

  return data;
}

function _prompt() {
  const data = _interopRequireDefault(require("../prompt"));

  _prompt = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async function userWantsToEjectWithoutUpgradingAsync() {
  const answer = await (0, _prompt().default)({
    type: 'confirm',
    name: 'ejectWithoutUpgrading',
    message: `We recommend upgrading to the latest SDK version before ejecting. SDK 37 introduces support for OTA updates and notifications in ejected projects, and includes many features that make ejecting your project easier. Would you like to continue ejecting anyways?`
  });
  return answer.ejectWithoutUpgrading;
}

async function action(projectDir, options) {
  let exp;

  try {
    exp = (0, _config().getConfig)(projectDir).exp;
  } catch (error) {
    console.log();
    console.log(_chalk().default.red(error.message));
    console.log();
    process.exit(1);
  } // Set EXPO_VIEW_DIR to universe/exponent to pull expo view code locally instead of from S3 for ExpoKit


  if (_xdl().Versions.lteSdkVersion(exp, '36.0.0')) {
    // Don't show a warning if we haven't released SDK 37 yet
    let latestReleasedVersion = await _xdl().Versions.newestReleasedSdkVersionAsync();

    if (_xdl().Versions.lteSdkVersion({
      sdkVersion: latestReleasedVersion.version
    }, '36.0.0')) {
      await LegacyEject().ejectAsync(projectDir, options);
    } else {
      if (options.force || (await userWantsToEjectWithoutUpgradingAsync())) {
        await LegacyEject().ejectAsync(projectDir, options);
      }
    }
  } else {
    await Eject().ejectAsync(projectDir, options);
  }
}

function _default(program) {
  program.command('eject [project-dir]').description('Creates Xcode and Android Studio projects for your app. Use this if you need to add custom native functionality.').option('--eject-method [type]', `Eject method to use. [Depreacted]: Ejecting to ExpoKit is not available on SDK >= 37 and not recommended for older SDK versions. We recommend updating to SDK >= 37 and ejecting to bare.`, value => value.toLowerCase()).option('-f --force', 'Will attempt to generate an iOS project even when the system is not running macOS. Unsafe and may fail.').asyncActionProjectDir(action);
}
