// expo-cli/build/commands/upgrade.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.maybeFormatSdkVersion = maybeFormatSdkVersion;
exports.getUpdatedDependenciesAsync = getUpdatedDependenciesAsync;
exports.getDependenciesFromBundledNativeModules = getDependenciesFromBundledNativeModules;
exports.upgradeAsync = upgradeAsync;
exports.default = _default;

function ConfigUtils() {
  const data = _interopRequireWildcard(require("@expo/config"));

  ConfigUtils = function () {
    return data;
  };

  return data;
}

function _jsonFile() {
  const data = _interopRequireDefault(require("@expo/json-file"));

  _jsonFile = function () {
    return data;
  };

  return data;
}

function PackageManager() {
  const data = _interopRequireWildcard(require("@expo/package-manager"));

  PackageManager = function () {
    return data;
  };

  return data;
}

function _xdl() {
  const data = require("@expo/xdl");

  _xdl = function () {
    return data;
  };

  return data;
}

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function () {
    return data;
  };

  return data;
}

function _commander() {
  const data = _interopRequireDefault(require("commander"));

  _commander = function () {
    return data;
  };

  return data;
}

function _pickBy() {
  const data = _interopRequireDefault(require("lodash/pickBy"));

  _pickBy = function () {
    return data;
  };

  return data;
}

function _difference() {
  const data = _interopRequireDefault(require("lodash/difference"));

  _difference = function () {
    return data;
  };

  return data;
}

function _semver() {
  const data = _interopRequireDefault(require("semver"));

  _semver = function () {
    return data;
  };

  return data;
}

function _ora() {
  const data = _interopRequireDefault(require("ora"));

  _ora = function () {
    return data;
  };

  return data;
}

function _terminalLink() {
  const data = _interopRequireDefault(require("terminal-link"));

  _terminalLink = function () {
    return data;
  };

  return data;
}

function _getenv() {
  const data = _interopRequireDefault(require("getenv"));

  _getenv = function () {
    return data;
  };

  return data;
}

function _log() {
  const data = _interopRequireDefault(require("../log"));

  _log = function () {
    return data;
  };

  return data;
}

function _prompt() {
  const data = _interopRequireDefault(require("../prompt"));

  _prompt = function () {
    return data;
  };

  return data;
}

function _ProjectUtils() {
  const data = require("./utils/ProjectUtils");

  _ProjectUtils = function () {
    return data;
  };

  return data;
}

function _maybeBailOnGitStatusAsync() {
  const data = _interopRequireDefault(require("./utils/maybeBailOnGitStatusAsync"));

  _maybeBailOnGitStatusAsync = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function logNewSection(title) {
  let spinner = (0, _ora().default)(_chalk().default.bold(title));
  spinner.start();
  return spinner;
}

function maybeFormatSdkVersion(sdkVersionString) {
  if (typeof sdkVersionString !== 'string' || sdkVersionString === 'UNVERSIONED') {
    return sdkVersionString;
  } // semver.valid type doesn't accept null, so need to ensure we pass a string


  return _semver().default.valid(_semver().default.coerce(sdkVersionString) || '');
}
/**
 * Produce a list of dependencies used by the project that need to be updated
 */


async function getUpdatedDependenciesAsync(projectRoot, workflow, targetSdkVersion) {
  // Get the updated version for any bundled modules
  const {
    exp,
    pkg
  } = ConfigUtils().getConfig(projectRoot);
  const bundledNativeModules = await _jsonFile().default.readAsync(ConfigUtils().resolveModule('expo/bundledNativeModules.json', projectRoot, exp)); // Smoosh regular and dev dependencies together for now

  const projectDependencies = { ...pkg.dependencies,
    ...pkg.devDependencies
  };
  return getDependenciesFromBundledNativeModules({
    projectDependencies,
    bundledNativeModules,
    sdkVersion: exp.sdkVersion,
    workflow,
    targetSdkVersion
  });
}

function getDependenciesFromBundledNativeModules({
  projectDependencies,
  bundledNativeModules,
  sdkVersion,
  workflow,
  targetSdkVersion
}) {
  const result = {};
  Object.keys(bundledNativeModules).forEach(name => {
    if (projectDependencies[name]) {
      result[name] = bundledNativeModules[name];
    }
  }); // If sdkVersion is known and jest-expo is used, then upgrade to the current sdk version
  // jest-expo is versioned with expo because jest-expo mocks out the native SDKs used it expo.

  if (sdkVersion && projectDependencies['jest-expo']) {
    result['jest-expo'] = `^${sdkVersion}`;
  }

  if (!targetSdkVersion) {
    _log().default.warn(`Supported react, react-native, and react-dom versions are unknown because we don't have version information for the target SDK, please update them manually.`);

    return result;
  } // Get the supported react/react-native/react-dom versions and other related packages


  if (workflow === 'managed' || projectDependencies['expokit']) {
    result['react-native'] = `https://github.com/expo/react-native/archive/${targetSdkVersion.expoReactNativeTag}.tar.gz`;
  } else {
    result['react-native'] = targetSdkVersion.facebookReactNativeVersion;
  } // React version apparently is optional in SDK version data


  if (targetSdkVersion.facebookReactVersion) {
    result['react'] = targetSdkVersion.facebookReactVersion; // react-dom version is always the same as the react version

    if (projectDependencies['react-dom']) {
      result['react-dom'] = targetSdkVersion.facebookReactVersion;
    }
  } // Update any related packages


  if (targetSdkVersion.relatedPackages) {
    Object.keys(targetSdkVersion.relatedPackages).forEach(name => {
      if (projectDependencies[name]) {
        result[name] = targetSdkVersion.relatedPackages[name];
      }
    });
  }

  return result;
}

async function makeBreakingChangesToConfigAsync(projectRoot, targetSdkVersionString) {
  var _rootConfig$expo, _rootConfig$expo$andr, _currentExp$androidNa;

  let step = logNewSection('Updating your app.json to account for breaking changes (if applicable)...');
  let {
    rootConfig
  } = await ConfigUtils().readConfigJsonAsync(projectRoot);
  const {
    exp: currentExp
  } = ConfigUtils().getConfig(projectRoot, {
    skipSDKVersionRequirement: true
  });

  try {
    switch (targetSdkVersionString) {
      case '37.0.0':
        if ((rootConfig === null || rootConfig === void 0 ? void 0 : (_rootConfig$expo = rootConfig.expo) === null || _rootConfig$expo === void 0 ? void 0 : (_rootConfig$expo$andr = _rootConfig$expo.androidNavigationBar) === null || _rootConfig$expo$andr === void 0 ? void 0 : _rootConfig$expo$andr.visible) !== undefined) {
          var _rootConfig$expo$andr2, _rootConfig$expo$andr3;

          if ((rootConfig === null || rootConfig === void 0 ? void 0 : (_rootConfig$expo$andr2 = rootConfig.expo.androidNavigationBar) === null || _rootConfig$expo$andr2 === void 0 ? void 0 : _rootConfig$expo$andr2.visible) === false) {
            step.succeed(`Updated "androidNavigationBar.visible" property in app.json to "leanback"...`);
            rootConfig.expo.androidNavigationBar.visible = 'leanback';
          } else if ((rootConfig === null || rootConfig === void 0 ? void 0 : (_rootConfig$expo$andr3 = rootConfig.expo.androidNavigationBar) === null || _rootConfig$expo$andr3 === void 0 ? void 0 : _rootConfig$expo$andr3.visible) === true) {
            var _rootConfig$expo$andr4;

            step.succeed(`Removed extraneous "androidNavigationBar.visible" property in app.json...`);
            rootConfig === null || rootConfig === void 0 ? true : (_rootConfig$expo$andr4 = rootConfig.expo.androidNavigationBar) === null || _rootConfig$expo$andr4 === void 0 ? true : delete _rootConfig$expo$andr4.visible;
          } else {
            // They had some invalid property for androidNavigationBar already...
            step.succeed('No additional changes necessary to app.json config.');
            return;
          }

          await ConfigUtils().writeConfigJsonAsync(projectRoot, rootConfig.expo);
        } else if ((currentExp === null || currentExp === void 0 ? void 0 : (_currentExp$androidNa = currentExp.androidNavigationBar) === null || _currentExp$androidNa === void 0 ? void 0 : _currentExp$androidNa.visible) !== undefined) {
          step.stopAndPersist({
            symbol: 'âš ï¸ ',
            text: _chalk().default.red(`Please manually update "androidNavigationBar.visible" according to ${(0, _terminalLink().default)('this documentation', 'https://docs.expo.io/workflow/configuration/#androidnavigationbar')}`)
          });
        } else {
          step.succeed('No additional changes necessary to app.json config.');
        }

        break;

      default:
        step.succeed('No additional changes necessary to app.json config.');
    }
  } catch (e) {
    step.fail(`Something went wrong when attempting to update app.json configuration: ${e.message}`);
  }
}

async function maybeBailOnUnsafeFunctionalityAsync(exp) {
  // Give people a chance to bail out if they're updating from a super old version because YMMV
  if (!_xdl().Versions.gteSdkVersion(exp, '33.0.0')) {
    if (_commander().default.nonInteractive) {
      _log().default.warn(`This command works best on SDK 33 and higher. Because the command is running in nonInteractive mode it'll continue regardless.`);

      return true;
    }

    const answer = await (0, _prompt().default)({
      type: 'confirm',
      name: 'attemptOldUpdate',
      message: `This command works best on SDK 33 and higher. We can try updating for you, but you will likely need to follow up with the instructions from https://docs.expo.io/workflow/upgrading-expo-sdk-walkthrough/. Continue anyways?`
    });

    if (!answer.attemptOldUpdate) {
      return true;
    }

    _log().default.newLine();
  }

  return false;
}

async function stopExpoServerAsync(projectRoot) {
  // Can't upgrade if Expo is running
  const status = await _xdl().Project.currentStatus(projectRoot);

  if (status === 'running') {
    await _xdl().Project.stopAsync(projectRoot);
    (0, _log().default)(_chalk().default.bold.underline('We found an existing expo-cli instance running for this project and closed it to continue.'));

    _log().default.addNewLineIfNone();
  }
}

async function shouldBailWhenUsingLatest(currentSdkVersionString, targetSdkVersionString) {
  // Maybe bail out early if people are trying to update to the current version
  if (targetSdkVersionString === currentSdkVersionString) {
    if (_commander().default.nonInteractive) {
      _log().default.warn(`You are already using the latest SDK version but the command will continue because nonInteractive is enabled.`);

      _log().default.newLine();

      return false;
    }

    const answer = await (0, _prompt().default)({
      type: 'confirm',
      name: 'attemptUpdateAgain',
      message: `You are already using the latest SDK version. Do you want to run the update anyways? This may be useful to ensure that all of your packages are set to the correct version.`
    });

    if (!answer.attemptUpdateAgain) {
      (0, _log().default)('Follow the Expo blog at https://blog.expo.io for new release information!');

      _log().default.newLine();

      return true;
    }

    _log().default.newLine();
  }

  return false;
}

async function shouldUpgradeSimulatorAsync() {
  // Check if we can, and probably should, upgrade the (ios) simulator
  if (_xdl().Simulator.isPlatformSupported()) {
    if (_commander().default.nonInteractive) {
      _log().default.warn(`Skipping attempt to upgrade the client app on iOS simulator.`);

      return false;
    }

    let answer = await (0, _prompt().default)({
      type: 'confirm',
      name: 'upgradeSimulator',
      message: 'Would you like to upgrade the Expo app in the iOS simulator?',
      default: false
    });

    _log().default.newLine();

    return answer.upgradeSimulator;
  }

  return false;
}

async function maybeUpgradeSimulatorAsync() {
  // Check if we can, and probably should, upgrade the (ios) simulator
  if (_xdl().Simulator.isPlatformSupported()) {
    if (await shouldUpgradeSimulatorAsync()) {
      let result = await _xdl().Simulator.upgradeExpoAsync();

      if (!result) {
        _log().default.error("The upgrade of your simulator didn't go as planned. You might have to reinstall it manually with expo client:install:ios.");
      }
    }
  }
}

async function shouldUpgradeEmulatorAsync() {
  // Check if we can, and probably should, upgrade the android client
  if (_xdl().Android.isPlatformSupported()) {
    if (_commander().default.nonInteractive) {
      _log().default.warn(`Skipping attempt to upgrade the Expo app on the Android emulator.`);

      return false;
    }

    const answer = await (0, _prompt().default)({
      type: 'confirm',
      name: 'upgradeAndroid',
      message: 'Would you like to upgrade the Expo app in the Android emulator?',
      default: false
    });

    _log().default.newLine();

    return answer.upgradeAndroid;
  }

  return false;
}

async function maybeUpgradeEmulatorAsync() {
  // Check if we can, and probably should, upgrade the android client
  if (await shouldUpgradeEmulatorAsync()) {
    const result = await _xdl().Android.upgradeExpoAsync();

    if (!result) {
      _log().default.error("The upgrade of your Android client didn't go as planned. You might have to reinstall it manually with expo client:install:android.");
    }

    _log().default.newLine();
  }
}

async function upgradeAsync({
  requestedSdkVersion,
  projectRoot,
  workflow
}, options) {
  let {
    exp,
    pkg
  } = await ConfigUtils().getConfig(projectRoot);
  if (await (0, _maybeBailOnGitStatusAsync().default)()) return;
  if (await maybeBailOnUnsafeFunctionalityAsync(exp)) return;
  await stopExpoServerAsync(projectRoot);
  let currentSdkVersionString = exp.sdkVersion;
  let sdkVersions = await _xdl().Versions.releasedSdkVersionsAsync();
  let latestSdkVersion = await _xdl().Versions.newestReleasedSdkVersionAsync();
  let latestSdkVersionString = latestSdkVersion.version;
  let targetSdkVersionString = maybeFormatSdkVersion(requestedSdkVersion) || latestSdkVersion.version;
  let targetSdkVersion = sdkVersions[targetSdkVersionString]; // Maybe bail out early if people are trying to update to the current version

  if (await shouldBailWhenUsingLatest(currentSdkVersionString, targetSdkVersionString)) return;
  const platforms = exp.platforms || [];

  if (targetSdkVersionString === latestSdkVersionString && currentSdkVersionString !== targetSdkVersionString && !_commander().default.nonInteractive) {
    let answer = await (0, _prompt().default)({
      type: 'confirm',
      name: 'updateToLatestSdkVersion',
      message: `You are currently using SDK ${currentSdkVersionString}. Would you like to update to the latest version, ${latestSdkVersion.version}?`
    });

    _log().default.newLine();

    if (!answer.updateToLatestSdkVersion) {
      let sdkVersionStringOptions = Object.keys(sdkVersions).filter(v => _semver().default.lte('33.0.0', v) && !_xdl().Versions.gteSdkVersion(exp, v));
      let {
        selectedSdkVersionString
      } = await (0, _prompt().default)({
        type: 'list',
        name: 'selectedSdkVersionString',
        message: 'Choose a SDK version to upgrade to:',
        pageSize: 20,
        choices: sdkVersionStringOptions.map(sdkVersionString => ({
          value: sdkVersionString,
          name: _chalk().default.bold(sdkVersionString)
        }))
      }); // This has to exist because it's based on keys already present in sdkVersions

      targetSdkVersion = sdkVersions[selectedSdkVersionString];
      targetSdkVersionString = selectedSdkVersionString;

      _log().default.newLine();
    } // Check if we can, and probably should, upgrade the (ios) simulator


    if (platforms.includes('ios')) {
      await maybeUpgradeSimulatorAsync();
    } // Check if we can, and probably should, upgrade the android client


    if (platforms.includes('android')) {
      await maybeUpgradeEmulatorAsync();
    }
  } else if (!targetSdkVersion) {
    if (_commander().default.nonInteractive) {
      _log().default.warn(`You provided the target SDK version value of ${targetSdkVersionString} which does not seem to exist. Upgrading anyways because the command is running in nonInteractive mode.`);
    } else {
      // If they provide an apparently unsupported sdk version then let people try
      // anyways, maybe we want to use this for testing alpha versions or
      // something...
      let answer = await (0, _prompt().default)({
        type: 'confirm',
        name: 'attemptUnknownUpdate',
        message: `You provided the target SDK version value of ${targetSdkVersionString} which does not seem to exist. But hey, I'm just a program, what do I know. Do you want to try to upgrade to it anyways?`
      });

      if (!answer.attemptUnknownUpdate) {
        return;
      }
    }
  }

  let packageManager = PackageManager().createForProject(projectRoot, {
    npm: options.npm,
    yarn: options.yarn,
    log: _log().default,
    silent: _getenv().default.boolish('EXPO_DEBUG', true)
  });

  _log().default.addNewLineIfNone();

  const expoPackageToInstall = `expo@^${targetSdkVersionString}`;
  let installingPackageStep = logNewSection(`Installing the ${expoPackageToInstall} package...`);

  _log().default.addNewLineIfNone();

  try {
    await packageManager.addAsync(expoPackageToInstall);
  } catch (e) {
    installingPackageStep.fail(`Failed to install expo package with error: ${e.message}`);
  } finally {
    installingPackageStep.succeed(`Installed ${expoPackageToInstall}`);
  } // Remove sdkVersion from app.json


  let removingSdkVersionStep = logNewSection('Validating configuration.');

  try {
    const {
      rootConfig
    } = await ConfigUtils().readConfigJsonAsync(projectRoot);

    if (rootConfig.expo.sdkVersion && rootConfig.expo.sdkVersion !== 'UNVERSIONED') {
      _log().default.addNewLineIfNone();

      await ConfigUtils().writeConfigJsonAsync(projectRoot, {
        sdkVersion: undefined
      });
      removingSdkVersionStep.succeed('Removed deprecated sdkVersion field from app.json.');
    } else {
      removingSdkVersionStep.succeed('Validated configuration.');
    }
  } catch (_) {
    removingSdkVersionStep.fail('Unable to validate configuration.');
  } // Evaluate project config (app.config.js)


  const {
    exp: currentExp
  } = ConfigUtils().getConfig(projectRoot);

  if (!_xdl().Versions.gteSdkVersion(currentExp, targetSdkVersionString) && currentExp.sdkVersion !== 'UNVERSIONED') {
    _log().default.addNewLineIfNone();

    removingSdkVersionStep.warn('Please manually delete the sdkVersion field in your project app.config file, it is deprecated.');
  }

  await makeBreakingChangesToConfigAsync(projectRoot, targetSdkVersionString);
  let updatingPackagesStep = logNewSection('Updating packages to compatible versions (where known).');

  _log().default.addNewLineIfNone(); // Get all updated packages


  let updates = await getUpdatedDependenciesAsync(projectRoot, workflow, targetSdkVersion); // Split updated packages by dependencies and devDependencies

  let devDependencies = (0, _pickBy().default)(updates, (_version, name) => {
    var _pkg$devDependencies;

    return (_pkg$devDependencies = pkg.devDependencies) === null || _pkg$devDependencies === void 0 ? void 0 : _pkg$devDependencies[name];
  });
  let devDependenciesAsStringArray = Object.keys(devDependencies).map(name => `${name}@${updates[name]}`);
  let dependencies = (0, _pickBy().default)(updates, (_version, name) => {
    var _pkg$dependencies;

    return (_pkg$dependencies = pkg.dependencies) === null || _pkg$dependencies === void 0 ? void 0 : _pkg$dependencies[name];
  });
  let dependenciesAsStringArray = Object.keys(dependencies).map(name => `${name}@${updates[name]}`); // Install dev dependencies

  if (devDependenciesAsStringArray.length) {
    try {
      await packageManager.addDevAsync(...devDependenciesAsStringArray);
    } catch (e) {
      updatingPackagesStep.fail(`Failed to upgrade JavaScript devDependencies: ${devDependenciesAsStringArray.join(' ')}`);
    }
  } // Install dependencies


  if (dependenciesAsStringArray.length) {
    try {
      await packageManager.addAsync(...dependenciesAsStringArray);
    } catch (e) {
      updatingPackagesStep.fail(`Failed to upgrade JavaScript dependencies: ${dependenciesAsStringArray.join(' ')}`);
    }
  }

  updatingPackagesStep.succeed('Updated known packages to compatible versions.'); // Remove package-lock.json and node_modules if using npm instead of yarn. See the function
  // for more information on why.

  await maybeCleanNpmStateAsync(packageManager);
  let clearingCacheStep = logNewSection('Clearing the packager cache.');

  try {
    await _xdl().Project.startReactNativeServerAsync(projectRoot, {
      reset: true,
      nonPersistent: true
    });
  } catch (e) {
    clearingCacheStep.fail(`Failed to clear packager cache with error: ${e.message}`);
  } finally {
    try {
      // Ensure that we at least attempt to stop the server even if it failed to clear the cache
      // It was pointed out to me that "Connecting to Metro bundler failed." could occur which would lead
      // to the upgrade command not exiting upon completion because, I believe, the server remained open.
      await _xdl().Project.stopReactNativeServerAsync(projectRoot);
    } catch (_unused) {}

    clearingCacheStep.succeed('Cleared packager cache.');
  }

  _log().default.newLine();

  (0, _log().default)(_chalk().default.bold.green(`ðŸ‘ Automated upgrade steps complete.`));
  (0, _log().default)(_chalk().default.bold.grey(`...but this doesn't mean everything is done yet!`));

  _log().default.newLine(); // List packages that were updated


  (0, _log().default)(_chalk().default.bold(`âœ… The following packages were updated:`));
  (0, _log().default)(_chalk().default.grey.bold([...Object.keys(updates), ...['expo']].join(', ')));

  _log().default.addNewLineIfNone(); // List packages that were not updated


  let allDependencies = { ...pkg.dependencies,
    ...pkg.devDependencies
  };
  let untouchedDependencies = (0, _difference().default)(Object.keys(allDependencies), [...Object.keys(updates), 'expo']);

  if (untouchedDependencies.length) {
    _log().default.addNewLineIfNone();

    (0, _log().default)(_chalk().default.bold(`ðŸš¨ The following packages were ${_chalk().default.underline('not')} updated. You should check the READMEs for those repositories to determine what version is compatible with your new set of packages:`));
    (0, _log().default)(_chalk().default.grey.bold(untouchedDependencies.join(', ')));

    _log().default.addNewLineIfNone();
  } // Add some basic additional instructions for bare workflow


  if (workflow === 'bare') {
    _log().default.addNewLineIfNone();

    (0, _log().default)(_chalk().default.bold(`It will be necessary to re-build your native projects to compile the updated dependencies. You will need to run ${_chalk().default.grey('pod install')} in your ios directory before re-building the iOS project.`));

    _log().default.addNewLineIfNone();
  }

  if (targetSdkVersion && targetSdkVersion.releaseNoteUrl) {
    (0, _log().default)(`Please refer to the release notes for information on any further required steps to update and information about breaking changes:`);
    (0, _log().default)(_chalk().default.bold(targetSdkVersion.releaseNoteUrl));
  } else {
    _log().default.gray(`Unable to find release notes for ${targetSdkVersionString}, please try to find them on https://blog.expo.io to learn more about other potentially important upgrade steps and breaking changes.`);
  }

  let skippedSdkVersions = (0, _pickBy().default)(sdkVersions, (_data, sdkVersionString) => {
    return _semver().default.lt(sdkVersionString, targetSdkVersionString) && _semver().default.gt(sdkVersionString, currentSdkVersionString);
  });
  let skippedSdkVersionKeys = Object.keys(skippedSdkVersions);

  if (skippedSdkVersionKeys.length) {
    _log().default.newLine();

    let releaseNotesUrls = Object.values(skippedSdkVersions).map(data => data.releaseNoteUrl).filter(releaseNoteUrl => releaseNoteUrl).reverse();

    if (releaseNotesUrls.length === 1) {
      (0, _log().default)(`You should also look at the breaking changes from a release that you skipped:`);
      (0, _log().default)(_chalk().default.bold(`- ${releaseNotesUrls[0]}`));
    } else {
      (0, _log().default)(`In addition to the most recent release notes, you should go over the breaking changes from skipped releases:`);
      releaseNotesUrls.forEach(url => {
        (0, _log().default)(_chalk().default.bold(`- ${url}`));
      });
    }
  }
}

async function maybeCleanNpmStateAsync(packageManager) {
  // We don't trust npm to properly handle deduping dependencies so we need to
  // clear the lockfile and node_modules.
  // https://forums.expo.io/t/sdk-37-unrecognized-font-family/35201
  // https://twitter.com/geoffreynyaga/status/1246170581109743617
  if (packageManager instanceof PackageManager().NpmPackageManager) {
    let cleaningNpmStateStep = logNewSection('Removing package-lock.json and deleting node_modules.');
    let shouldInstallNodeModules = true;

    try {
      await packageManager.removeLockfileAsync();
      await packageManager.cleanAsync();
      cleaningNpmStateStep.succeed('Removed package-lock.json and deleted node_modules.');
    } catch (_unused2) {
      shouldInstallNodeModules = false;
      cleaningNpmStateStep.fail('Unable to remove package-lock.json and delete node_modules. We recommend doing this to ensure that the upgrade goes smoothly when using npm instead of yarn.');
    }

    if (shouldInstallNodeModules) {
      let reinstallingNodeModulesStep = logNewSection('Installing node_modules and rebuilding package-lock.json.');

      try {
        await packageManager.installAsync();
        reinstallingNodeModulesStep.succeed('Installed node_modules and rebuilt package-lock.json.');
      } catch (_unused3) {
        reinstallingNodeModulesStep.fail('Running npm install failed. Please check npm-error.log for more information.');
      }
    }
  }
}

function _default(program) {
  program.command('upgrade [targetSdkVersion]').alias('update').option('--npm', 'Use npm to install dependencies. (default when package-lock.json exists)').option('--yarn', 'Use Yarn to install dependencies. (default when yarn.lock exists)').description('Upgrades your project dependencies and app.json and to the given SDK version').asyncAction(async (requestedSdkVersion, options) => {
    const {
      projectRoot,
      workflow
    } = await (0, _ProjectUtils().findProjectRootAsync)(process.cwd());
    await upgradeAsync({
      requestedSdkVersion,
      projectRoot,
      workflow
    }, options);
  });
}
