// @expo/schemer/build/Error.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const es6_error_1 = __importDefault(require("es6-error"));
class SchemerError extends es6_error_1.default {
    constructor(errors) {
        const message = errors.map(e => e.message).join('\n');
        super(message);
        this.errors = errors;
    }
}
exports.SchemerError = SchemerError;
class ValidationError extends es6_error_1.default {
    constructor({ errorCode, fieldPath, message, data, meta, }) {
        super(message);
        this.errorCode = errorCode;
        this.fieldPath = fieldPath;
        this.message = message;
        this.data = data;
        this.meta = meta;
    }
}
exports.ValidationError = ValidationError;
exports.ErrorCodes = {
    SCHEMA_VALIDATION_ERROR: 'SCHEMA_VALIDATION_ERROR',
    SCHEMA_ADDITIONAL_PROPERTY: 'SCHEMA_ADDITIONAL_PROPERTY',
    SCHEMA_MISSING_REQUIRED_PROPERTY: 'SCHEMA_MISSING_REQUIRED_PROPERTY',
    SCHEMA_INVALID_PATTERN: 'SCHEMA_INVALID_PATTERN',
    INVALID_ASSET_URI: 'INVALID_ASSET_URI',
    INVALID_DIMENSIONS: 'INVALID_DIMENSIONS',
    INVALID_CONTENT_TYPE: 'INVALID_CONTENT_TYPE',
    NOT_SQUARE: 'NOT_SQUARE',
};
