// @expo/build-tools/dist/platforms/ios/credentials.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const v4_1 = __importDefault(require("uuid/v4"));
const distributionCertificateUtils = __importStar(require("./utils/distributionCertificate"));
const keychain_1 = __importDefault(require("./utils/keychain"));
const provisioningProfile_1 = __importDefault(require("./utils/provisioningProfile"));
class iOSCredentialsManager {
    constructor(logger, job) {
        this.cleanedUp = false;
        this.logger = logger;
        this.job = job;
        this.distributionCertificatePath = path_1.default.join(os_1.default.tmpdir(), `${v4_1.default()}.p12`);
    }
    async prepare() {
        const { secrets } = this.job;
        this.logger.info('Getting distribution certificate fingerprint');
        const certificateFingerprint = distributionCertificateUtils.getFingerprint(secrets.distributionCertificate);
        this.logger.info(`Writing distribution certificate to ${this.distributionCertificatePath}`);
        await fs_extra_1.default.writeFile(this.distributionCertificatePath, Buffer.from(secrets.distributionCertificate.dataBase64, 'base64'));
        try {
            this.logger.info('Creating keychain');
            this.keychain = new keychain_1.default(this.logger);
            await this.keychain.create();
            this.logger.info('Importing distribution certificate into the keychain');
            await this.keychain.importCertificate(this.distributionCertificatePath, secrets.distributionCertificate.password);
            this.logger.info('Initializing provisioning profile');
            this.provisioningProfile = new provisioningProfile_1.default(this.logger, Buffer.from(secrets.provisioningProfileBase64, 'base64'));
            await this.provisioningProfile.init();
            this.logger.info('Validating whether distribution certificate has been imported successfully');
            await this.keychain.ensureCertificateImported(this.provisioningProfile.data.teamId, certificateFingerprint);
            this.logger.info('Verifying whether the distribution certificate and provisioning profile match');
            this.provisioningProfile.verifyCertificate(certificateFingerprint);
            return {
                bundleIdentifier: this.provisioningProfile.data.bundleIdentifier,
                keychainPath: this.keychain.data.path,
                provisioningProfileUUID: this.provisioningProfile.data.uuid,
                teamId: this.provisioningProfile.data.teamId,
            };
        }
        catch (err) {
            await this.cleanUp();
            throw err;
        }
    }
    async cleanUp() {
        if (this.cleanedUp) {
            return;
        }
        if (this.keychain) {
            await this.keychain.destroy();
        }
        if (this.provisioningProfile) {
            await this.provisioningProfile.destroy();
        }
        this.cleanedUp = true;
    }
}
exports.iOSCredentialsManager = iOSCredentialsManager;
