// @expo/build-tools/dist/buildConfig/schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("@hapi/joi"));
const constants_1 = require("../constants");
const KeystoreInfoSchema = joi_1.default.object({
    keystorePath: joi_1.default.string().required(),
    keystorePassword: joi_1.default.string().required(),
    keyAlias: joi_1.default.string().required(),
    keyPassword: joi_1.default.string().required(),
});
const iOSCredentialsSchema = joi_1.default.object({
    provisioningProfilePath: joi_1.default.string().required(),
    distributionCertificate: joi_1.default.object({
        path: joi_1.default.string().required(),
        password: joi_1.default.string().required(),
    }).required(),
});
const SecretsSchema = joi_1.default.object({
    secretFiles: joi_1.default.array().items(joi_1.default.string()),
});
const GenericBuildConfigSchema = joi_1.default.object({
    type: joi_1.default.string().valid(constants_1.BuildType.Generic),
    android: SecretsSchema,
    ios: iOSCredentialsSchema.concat(SecretsSchema),
});
const ManagedBuildConfigSchema = joi_1.default.object({
    type: joi_1.default.string().valid(constants_1.BuildType.Managed),
    android: joi_1.default.object({ keystore: KeystoreInfoSchema.required() }),
    ios: iOSCredentialsSchema,
});
const BuildConfigSchema = joi_1.default.object({
    type: joi_1.default.string().valid(constants_1.BuildType.Generic, constants_1.BuildType.Managed).required(),
})
    .when(joi_1.default.object({ type: constants_1.BuildType.Generic }).unknown(), { then: GenericBuildConfigSchema })
    .when(joi_1.default.object({ type: constants_1.BuildType.Managed }).unknown(), { then: ManagedBuildConfigSchema });
exports.default = BuildConfigSchema;
