// expo-cli/build/credentials/views/SetupIosProvisioningProfile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SetupIosProvisioningProfile = void 0;

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function () {
    return data;
  };

  return data;
}

function _log() {
  const data = _interopRequireDefault(require("../../log"));

  _log = function () {
    return data;
  };

  return data;
}

function iosProfileView() {
  const data = _interopRequireWildcard(require("./IosProvisioningProfile"));

  iosProfileView = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class SetupIosProvisioningProfile {
  constructor(options) {
    var _options$nonInteracti;

    _defineProperty(this, "_experienceName", void 0);

    _defineProperty(this, "_bundleIdentifier", void 0);

    _defineProperty(this, "_distCert", void 0);

    _defineProperty(this, "_nonInteractive", void 0);

    const {
      experienceName,
      bundleIdentifier,
      distCert
    } = options;
    this._experienceName = experienceName;
    this._bundleIdentifier = bundleIdentifier;
    this._distCert = distCert;
    this._nonInteractive = (_options$nonInteracti = options.nonInteractive) !== null && _options$nonInteracti !== void 0 ? _options$nonInteracti : false;
  }

  async open(ctx) {
    if (!ctx.user) {
      throw new Error(`This workflow requires you to be logged in.`);
    }

    const appCredentials = await ctx.ios.getAppCredentials(this._experienceName, this._bundleIdentifier); // Try to use the profile we have on file first

    const configuredProfile = await ctx.ios.getProvisioningProfile(this._experienceName, this._bundleIdentifier); // We dont have a profile on expo servers or
    // The configured profile is associated with some other dist cert

    const configuredWithSameDistCert = appCredentials.distCredentialsId === this._distCert.id;

    if (!configuredProfile || !configuredWithSameDistCert) {
      return new (iosProfileView().CreateOrReuseProvisioningProfile)({
        experienceName: this._experienceName,
        bundleIdentifier: this._bundleIdentifier,
        distCert: this._distCert,
        nonInteractive: this._nonInteractive
      });
    }

    if (!ctx.hasAppleCtx()) {
      const isValid = await iosProfileView().validateProfileWithoutApple(configuredProfile, this._distCert, this._bundleIdentifier);

      if (!isValid) {
        throw new Error(`The provisioning profile we have on file is no longer valid.`);
      }

      return null;
    } // User uploaded profiles dont have ids - do best effort validation here


    if (!configuredProfile.provisioningProfileId) {
      (0, _log().default)(_chalk().default.yellow("The provisioning profile we have on file cannot be validated on Apple's servers."));
      const isValid = await iosProfileView().validateProfileWithoutApple(configuredProfile, this._distCert, this._bundleIdentifier);

      if (!isValid) {
        return new (iosProfileView().CreateOrReuseProvisioningProfile)({
          experienceName: this._experienceName,
          bundleIdentifier: this._bundleIdentifier,
          distCert: this._distCert,
          nonInteractive: this._nonInteractive
        });
      }

      return null;
    }

    const profileFromApple = await iosProfileView().getAppleInfo(ctx.appleCtx, this._bundleIdentifier, configuredProfile); // Profile can't be found on Apple servers

    if (!profileFromApple) {
      return new (iosProfileView().CreateOrReuseProvisioningProfile)({
        experienceName: this._experienceName,
        bundleIdentifier: this._bundleIdentifier,
        distCert: this._distCert,
        nonInteractive: this._nonInteractive
      });
    }

    await iosProfileView().configureAndUpdateProvisioningProfile(ctx, this._experienceName, this._bundleIdentifier, this._distCert, profileFromApple);
    return null;
  }

}

exports.SetupIosProvisioningProfile = SetupIosProvisioningProfile;
