// expo-cli/build/credentials/views/SetupIosPush.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SetupIosPush = void 0;

function _inquirer() {
  const data = require("inquirer");

  _inquirer = function () {
    return data;
  };

  return data;
}

function iosPushView() {
  const data = _interopRequireWildcard(require("./IosPushCredentials"));

  iosPushView = function () {
    return data;
  };

  return data;
}

function _log() {
  const data = _interopRequireDefault(require("../../log"));

  _log = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class SetupIosPush {
  constructor(options) {
    var _options$nonInteracti;

    _defineProperty(this, "_experienceName", void 0);

    _defineProperty(this, "_bundleIdentifier", void 0);

    _defineProperty(this, "_nonInteractive", void 0);

    const {
      experienceName,
      bundleIdentifier
    } = options;
    this._experienceName = experienceName;
    this._bundleIdentifier = bundleIdentifier;
    this._nonInteractive = (_options$nonInteracti = options.nonInteractive) !== null && _options$nonInteracti !== void 0 ? _options$nonInteracti : false;
  }

  async open(ctx) {
    var _appCredentials$crede, _appCredentials$crede2, _appCredentials$crede3;

    if (!ctx.user) {
      throw new Error(`This workflow requires you to be logged in.`);
    } // TODO: Remove this on Nov 2020 when Apple no longer accepts deprecated push certs


    const appCredentials = await ctx.ios.getAppCredentials(this._experienceName, this._bundleIdentifier);
    const deprecatedPushId = appCredentials === null || appCredentials === void 0 ? void 0 : (_appCredentials$crede = appCredentials.credentials) === null || _appCredentials$crede === void 0 ? void 0 : _appCredentials$crede.pushId;
    const deprecatedPushP12 = appCredentials === null || appCredentials === void 0 ? void 0 : (_appCredentials$crede2 = appCredentials.credentials) === null || _appCredentials$crede2 === void 0 ? void 0 : _appCredentials$crede2.pushP12;
    const deprecatedPushPassword = appCredentials === null || appCredentials === void 0 ? void 0 : (_appCredentials$crede3 = appCredentials.credentials) === null || _appCredentials$crede3 === void 0 ? void 0 : _appCredentials$crede3.pushPassword;

    if (deprecatedPushId && deprecatedPushP12 && deprecatedPushPassword) {
      const confirmQuestion = {
        type: 'confirm',
        name: 'confirm',
        message: `We've detected legacy Push Certificates on file. Would you like to upgrade to the newer standard?`,
        pageSize: Infinity
      };
      const {
        confirm
      } = await (0, _inquirer().prompt)(confirmQuestion);

      if (!confirm) {
        (0, _log().default)(`Using Deprecated Push Cert: ${deprecatedPushId} on file`);
        return null;
      }
    }

    const configuredPushKey = await ctx.ios.getPushKey(this._experienceName, this._bundleIdentifier);

    if (configuredPushKey) {
      // we dont need to setup if we have a valid push key on file
      const isValid = await iosPushView().validatePushKey(ctx, configuredPushKey);

      if (isValid) {
        return null;
      }
    }

    return new (iosPushView().CreateOrReusePushKey)({
      experienceName: this._experienceName,
      bundleIdentifier: this._bundleIdentifier,
      nonInteractive: this._nonInteractive
    });
  }

}

exports.SetupIosPush = SetupIosPush;
