// @expo/build-tools/dist/platforms/ios/managed/prepare.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const constants_1 = require("../../../constants");
async function prepareJob(config, projectUrl, projectDir) {
    const provisioningProfilePath = path_1.default.join(projectDir, config.provisioningProfilePath);
    const provisioningProfileData = await fs_extra_1.default.readFile(provisioningProfilePath, 'base64');
    const distributionCertificatePath = path_1.default.join(projectDir, config.distributionCertificate.path);
    const distributionCertificateData = await fs_extra_1.default.readFile(distributionCertificatePath, 'base64');
    const secrets = {
        provisioningProfileBase64: provisioningProfileData,
        distributionCertificate: {
            dataBase64: distributionCertificateData,
            password: config.distributionCertificate.password,
        },
    };
    const packageJson = await fs_extra_1.default.readFile(path_1.default.join(projectDir, 'package.json'), 'utf8');
    return {
        platform: constants_1.Platform.iOS,
        projectUrl,
        type: constants_1.BuildType.Managed,
        secrets,
        packageJson,
        manifest: 'TODO',
    };
}
exports.default = prepareJob;
