// @expo/build-tools/dist/platforms/android/managed/job.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("@hapi/joi"));
const constants_1 = require("../../../constants");
const baseJob_1 = require("../baseJob");
const AndroidKeystoreSchema = joi_1.default.object().keys({
    keystore: joi_1.default.object().keys({
        keystoreDataBase64: joi_1.default.string().required(),
        keystorePath: joi_1.default.string().required(),
        keystorePassword: joi_1.default.string().required(),
        keyAlias: joi_1.default.string().required(),
        keyPassword: joi_1.default.string().required(),
    }).required(),
});
const ManagedAndroidJobSchema = baseJob_1.BaseAndroidJobSchema.concat(joi_1.default.object().keys({
    type: joi_1.default.string().valid(constants_1.BuildType.Managed),
    packageJson: joi_1.default.object().unknown(true).required(),
    manifest: joi_1.default.object().unknown(true).required(),
    secrets: joi_1.default.object({ keystore: AndroidKeystoreSchema.required() }),
}));
exports.ManagedAndroidJobSchema = ManagedAndroidJobSchema;
