// @expo/dev-tools/build/server/graphql/Issues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
class Issues extends events_1.EventEmitter {
    constructor() {
        super(...arguments);
        this.issues = {};
    }
    addIssue(issueId, issue) {
        let newIssue = false;
        if (!this.issues[issueId]) {
            newIssue = true;
        }
        this.issues[issueId] = Object.assign(Object.assign({}, issue), { id: issueId });
        if (newIssue) {
            this.emit('ADDED', this.issues[issueId]);
        }
        else {
            this.emit('UPDATED', this.issues[issueId]);
        }
    }
    clearIssue(issueId) {
        const issue = this.issues[issueId];
        if (issue) {
            delete this.issues[issueId];
            this.emit('DELETED', issue);
        }
    }
    getIssueList() {
        return Object.keys(this.issues).map(key => ({
            cursor: key,
            node: this.issues[key],
        }));
    }
}
exports.default = Issues;
