// @expo/config/build/ios/utils/Xcodeproj.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-ignore
const xcode_1 = require("xcode");
// @ts-ignore
const pbxFile_1 = __importDefault(require("xcode/lib/pbxFile"));
const glob_1 = require("glob");
const path_1 = __importDefault(require("path"));
function getProjectName(projectRoot) {
    const sourceRoot = getSourceRoot(projectRoot);
    const sourceRootParts = sourceRoot.split(path_1.default.sep);
    return sourceRootParts[sourceRootParts.length - 1];
}
exports.getProjectName = getProjectName;
function getSourceRoot(projectRoot) {
    const paths = glob_1.sync('ios/*/AppDelegate.m', {
        absolute: true,
        cwd: projectRoot,
    });
    return path_1.default.dirname(paths[0]);
}
exports.getSourceRoot = getSourceRoot;
// TODO(brentvatne): I couldn't figure out how to do this with an existing
// higher level function exposed by the xcode library, but we should find out how to do
// that and replace this with it
function addFileToGroup(filepath, groupName, project) {
    let file = new pbxFile_1.default(filepath);
    file.uuid = project.generateUuid();
    file.fileRef = project.generateUuid();
    project.addToPbxFileReferenceSection(file);
    project.addToPbxBuildFileSection(file);
    project.addToPbxSourcesBuildPhase(file);
    let group = project.pbxGroupByName(groupName);
    if (!group) {
        throw Error(`Group by name ${groupName} not found!`);
    }
    group.children.push({ value: file.fileRef, comment: file.basename });
    return project;
}
exports.addFileToGroup = addFileToGroup;
/**
 * Get the pbxproj for the given path
 */
function getPbxproj(projectRoot) {
    const pbxprojPaths = glob_1.sync('ios/*/project.pbxproj', { absolute: true, cwd: projectRoot });
    const [pbxprojPath, ...otherPbxprojPaths] = pbxprojPaths;
    if (pbxprojPaths.length > 1) {
        console.warn(`Found multiple pbxproject files paths, using ${pbxprojPath}. Other paths ${JSON.stringify(otherPbxprojPaths)} ignored.`);
    }
    const project = xcode_1.project(pbxprojPath);
    project.parseSync();
    return project;
}
exports.getPbxproj = getPbxproj;
function removeComments([item]) {
    return !item.endsWith(`_comment`);
}
exports.removeComments = removeComments;
function isBuildConfig(input) {
    const { 1: { isa }, } = input;
    return isa === 'XCBuildConfiguration';
}
exports.isBuildConfig = isBuildConfig;
function removeTestHosts(input) {
    const { 1: { buildSettings }, } = input;
    return !buildSettings.TEST_HOST;
}
exports.removeTestHosts = removeTestHosts;
