// @expo/build-tools/dist/platforms/ios/utils/gymfile.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const template_file_1 = __importDefault(require("@expo/template-file"));
const TEMPLATE_FILE_PATH = path_1.default.join(__dirname, '../../../../templates/Gymfile.template');
async function createIfNotExists(credentials, { nativeProjectDirectory, outputName, outputDirectory }) {
    const outputFile = path_1.default.join(nativeProjectDirectory, 'Gymfile');
    if (await fs_extra_1.default.pathExists(outputFile)) {
        return false;
    }
    else {
        await template_file_1.default(TEMPLATE_FILE_PATH, {
            BUNDLE_IDENTIFIER: credentials.bundleIdentifier,
            PROVISIONING_PROFILE_UUID: credentials.provisioningProfileUUID,
            KEYCHAIN_PATH: credentials.keychainPath,
            OUTPUT_DIRECTORY: outputDirectory,
            OUTPUT_NAME: outputName,
        }, outputFile);
        return true;
    }
}
exports.createIfNotExists = createIfNotExists;
