// @expo/xdl/build/xdl.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Api", {
  enumerable: true,
  get: function () {
    return _Api().default;
  }
});
Object.defineProperty(exports, "ApiV2", {
  enumerable: true,
  get: function () {
    return _ApiV().default;
  }
});
Object.defineProperty(exports, "Config", {
  enumerable: true,
  get: function () {
    return _Config().default;
  }
});
Object.defineProperty(exports, "ErrorCode", {
  enumerable: true,
  get: function () {
    return _ErrorCode().ErrorCode;
  }
});
Object.defineProperty(exports, "FormData", {
  enumerable: true,
  get: function () {
    return _FormData().default;
  }
});
Object.defineProperty(exports, "Logger", {
  enumerable: true,
  get: function () {
    return _Logger().default;
  }
});
Object.defineProperty(exports, "LoggerDetach", {
  enumerable: true,
  get: function () {
    return _Logger2().default;
  }
});
Object.defineProperty(exports, "NotificationCode", {
  enumerable: true,
  get: function () {
    return _NotificationCode().default;
  }
});
Object.defineProperty(exports, "PackagerLogsStream", {
  enumerable: true,
  get: function () {
    return _PackagerLogsStream().default;
  }
});
Object.defineProperty(exports, "LogRecord", {
  enumerable: true,
  get: function () {
    return _PackagerLogsStream().LogRecord;
  }
});
Object.defineProperty(exports, "LogUpdater", {
  enumerable: true,
  get: function () {
    return _PackagerLogsStream().LogUpdater;
  }
});
Object.defineProperty(exports, "TurtleApi", {
  enumerable: true,
  get: function () {
    return _TurtleApi().default;
  }
});
Object.defineProperty(exports, "UserManager", {
  enumerable: true,
  get: function () {
    return _User().default;
  }
});
Object.defineProperty(exports, "RegistrationData", {
  enumerable: true,
  get: function () {
    return _User().RegistrationData;
  }
});
Object.defineProperty(exports, "User", {
  enumerable: true,
  get: function () {
    return _User().User;
  }
});
Object.defineProperty(exports, "UserSettings", {
  enumerable: true,
  get: function () {
    return _UserSettings().default;
  }
});
Object.defineProperty(exports, "XDLError", {
  enumerable: true,
  get: function () {
    return _XDLError().default;
  }
});
exports.Webpack = exports.Web = exports.Versions = exports.Utils = exports.UrlUtils = exports.UpdateVersions = exports.ThirdParty = exports.Simulator = exports.ProjectUtils = exports.ProjectSettings = exports.Project = exports.PKCS12Utils = exports.Modules = exports.ModuleVersion = exports.IosPlist = exports.IosKeychain = exports.IosIcons = exports.StandaloneBuild = exports.ImageUtils = exports.FsCache = exports.FileSystem = exports.ExpSchema = exports.Exp = exports.ExponentTools = exports.Env = exports.Doctor = exports.AndroidCredentials = exports.Credentials = exports.ConnectionStatus = exports.Webhooks = exports.Binaries = exports.Android = exports.Analytics = exports.IosCodeSigning = exports.IosShellApp = exports.IosPodsTools = exports.IosWorkspace = exports.IosIPABuilder = exports.Detach = exports.AndroidShellApp = void 0;

function _sourceMapSupport() {
  const data = require("source-map-support");

  _sourceMapSupport = function () {
    return data;
  };

  return data;
}

function Analytics() {
  const data = _interopRequireWildcard(require("./Analytics"));

  Analytics = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Analytics", {
  enumerable: true,
  get: function () {
    return Analytics();
  }
});

function Android() {
  const data = _interopRequireWildcard(require("./Android"));

  Android = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Android", {
  enumerable: true,
  get: function () {
    return Android();
  }
});

function _Api() {
  const data = _interopRequireDefault(require("./Api"));

  _Api = function () {
    return data;
  };

  return data;
}

function _ApiV() {
  const data = _interopRequireDefault(require("./ApiV2"));

  _ApiV = function () {
    return data;
  };

  return data;
}

function Binaries() {
  const data = _interopRequireWildcard(require("./Binaries"));

  Binaries = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Binaries", {
  enumerable: true,
  get: function () {
    return Binaries();
  }
});

function Webhooks() {
  const data = _interopRequireWildcard(require("./Webhooks"));

  Webhooks = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Webhooks", {
  enumerable: true,
  get: function () {
    return Webhooks();
  }
});

function _Config() {
  const data = _interopRequireDefault(require("./Config"));

  _Config = function () {
    return data;
  };

  return data;
}

function ConnectionStatus() {
  const data = _interopRequireWildcard(require("./ConnectionStatus"));

  ConnectionStatus = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ConnectionStatus", {
  enumerable: true,
  get: function () {
    return ConnectionStatus();
  }
});

function Credentials() {
  const data = _interopRequireWildcard(require("./credentials/Credentials"));

  Credentials = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Credentials", {
  enumerable: true,
  get: function () {
    return Credentials();
  }
});

function AndroidCredentials() {
  const data = _interopRequireWildcard(require("./credentials/AndroidCredentials"));

  AndroidCredentials = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "AndroidCredentials", {
  enumerable: true,
  get: function () {
    return AndroidCredentials();
  }
});

function Doctor() {
  const data = _interopRequireWildcard(require("./project/Doctor"));

  Doctor = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Doctor", {
  enumerable: true,
  get: function () {
    return Doctor();
  }
});

function Env() {
  const data = _interopRequireWildcard(require("./Env"));

  Env = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Env", {
  enumerable: true,
  get: function () {
    return Env();
  }
});

function ExponentTools() {
  const data = _interopRequireWildcard(require("./detach/ExponentTools"));

  ExponentTools = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ExponentTools", {
  enumerable: true,
  get: function () {
    return ExponentTools();
  }
});

function _ErrorCode() {
  const data = require("./ErrorCode");

  _ErrorCode = function () {
    return data;
  };

  return data;
}

function Exp() {
  const data = _interopRequireWildcard(require("./Exp"));

  Exp = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Exp", {
  enumerable: true,
  get: function () {
    return Exp();
  }
});

function ExpSchema() {
  const data = _interopRequireWildcard(require("./project/ExpSchema"));

  ExpSchema = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ExpSchema", {
  enumerable: true,
  get: function () {
    return ExpSchema();
  }
});

function FileSystem() {
  const data = _interopRequireWildcard(require("./FileSystem"));

  FileSystem = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "FileSystem", {
  enumerable: true,
  get: function () {
    return FileSystem();
  }
});

function _FormData() {
  const data = _interopRequireDefault(require("./tools/FormData"));

  _FormData = function () {
    return data;
  };

  return data;
}

function FsCache() {
  const data = _interopRequireWildcard(require("./tools/FsCache"));

  FsCache = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "FsCache", {
  enumerable: true,
  get: function () {
    return FsCache();
  }
});

function ImageUtils() {
  const data = _interopRequireWildcard(require("./tools/ImageUtils"));

  ImageUtils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ImageUtils", {
  enumerable: true,
  get: function () {
    return ImageUtils();
  }
});

function StandaloneBuild() {
  const data = _interopRequireWildcard(require("./StandaloneBuild"));

  StandaloneBuild = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "StandaloneBuild", {
  enumerable: true,
  get: function () {
    return StandaloneBuild();
  }
});

function IosIcons() {
  const data = _interopRequireWildcard(require("./detach/IosIcons"));

  IosIcons = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "IosIcons", {
  enumerable: true,
  get: function () {
    return IosIcons();
  }
});

function IosKeychain() {
  const data = _interopRequireWildcard(require("./detach/IosKeychain"));

  IosKeychain = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "IosKeychain", {
  enumerable: true,
  get: function () {
    return IosKeychain();
  }
});

function IosPlist() {
  const data = _interopRequireWildcard(require("./detach/IosPlist"));

  IosPlist = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "IosPlist", {
  enumerable: true,
  get: function () {
    return IosPlist();
  }
});

function _Logger() {
  const data = _interopRequireDefault(require("./Logger"));

  _Logger = function () {
    return data;
  };

  return data;
}

function _Logger2() {
  const data = _interopRequireDefault(require("./detach/Logger"));

  _Logger2 = function () {
    return data;
  };

  return data;
}

function ModuleVersion() {
  const data = _interopRequireWildcard(require("./tools/ModuleVersion"));

  ModuleVersion = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ModuleVersion", {
  enumerable: true,
  get: function () {
    return ModuleVersion();
  }
});

function Modules() {
  const data = _interopRequireWildcard(require("./modules/Modules"));

  Modules = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Modules", {
  enumerable: true,
  get: function () {
    return Modules();
  }
});

function _NotificationCode() {
  const data = _interopRequireDefault(require("./NotificationCode"));

  _NotificationCode = function () {
    return data;
  };

  return data;
}

function _PackagerLogsStream() {
  const data = _interopRequireWildcard(require("./logs/PackagerLogsStream"));

  _PackagerLogsStream = function () {
    return data;
  };

  return data;
}

function PKCS12Utils() {
  const data = _interopRequireWildcard(require("./detach/PKCS12Utils"));

  PKCS12Utils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "PKCS12Utils", {
  enumerable: true,
  get: function () {
    return PKCS12Utils();
  }
});

function Project() {
  const data = _interopRequireWildcard(require("./Project"));

  Project = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Project", {
  enumerable: true,
  get: function () {
    return Project();
  }
});

function ProjectSettings() {
  const data = _interopRequireWildcard(require("./ProjectSettings"));

  ProjectSettings = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ProjectSettings", {
  enumerable: true,
  get: function () {
    return ProjectSettings();
  }
});

function ProjectUtils() {
  const data = _interopRequireWildcard(require("./project/ProjectUtils"));

  ProjectUtils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ProjectUtils", {
  enumerable: true,
  get: function () {
    return ProjectUtils();
  }
});

function Simulator() {
  const data = _interopRequireWildcard(require("./Simulator"));

  Simulator = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Simulator", {
  enumerable: true,
  get: function () {
    return Simulator();
  }
});

function ThirdParty() {
  const data = _interopRequireWildcard(require("./ThirdParty"));

  ThirdParty = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ThirdParty", {
  enumerable: true,
  get: function () {
    return ThirdParty();
  }
});

function _TurtleApi() {
  const data = _interopRequireDefault(require("./TurtleApi"));

  _TurtleApi = function () {
    return data;
  };

  return data;
}

function UpdateVersions() {
  const data = _interopRequireWildcard(require("./tools/UpdateVersions"));

  UpdateVersions = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "UpdateVersions", {
  enumerable: true,
  get: function () {
    return UpdateVersions();
  }
});

function UrlUtils() {
  const data = _interopRequireWildcard(require("./UrlUtils"));

  UrlUtils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "UrlUtils", {
  enumerable: true,
  get: function () {
    return UrlUtils();
  }
});

function _User() {
  const data = _interopRequireWildcard(require("./User"));

  _User = function () {
    return data;
  };

  return data;
}

function _UserSettings() {
  const data = _interopRequireDefault(require("./UserSettings"));

  _UserSettings = function () {
    return data;
  };

  return data;
}

function Utils() {
  const data = _interopRequireWildcard(require("./Utils"));

  Utils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Utils", {
  enumerable: true,
  get: function () {
    return Utils();
  }
});

function Versions() {
  const data = _interopRequireWildcard(require("./Versions"));

  Versions = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Versions", {
  enumerable: true,
  get: function () {
    return Versions();
  }
});

function Web() {
  const data = _interopRequireWildcard(require("./Web"));

  Web = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Web", {
  enumerable: true,
  get: function () {
    return Web();
  }
});

function Webpack() {
  const data = _interopRequireWildcard(require("./Webpack"));

  Webpack = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Webpack", {
  enumerable: true,
  get: function () {
    return Webpack();
  }
});

function _XDLError() {
  const data = _interopRequireDefault(require("./XDLError"));

  _XDLError = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

if (process.env.NODE_ENV !== 'production') {
  (0, _sourceMapSupport().install)();
}

const AndroidShellApp = require('./detach/AndroidShellApp.js');

exports.AndroidShellApp = AndroidShellApp;

const Detach = require('./detach/Detach.js');

exports.Detach = Detach;

const IosIPABuilder = require('./detach/IosIPABuilder.js').default;

exports.IosIPABuilder = IosIPABuilder;

const IosWorkspace = require('./detach/IosWorkspace');

exports.IosWorkspace = IosWorkspace;

const IosPodsTools = require('./detach/IosPodsTools.js');

exports.IosPodsTools = IosPodsTools;

const IosShellApp = require('./detach/IosShellApp.js');

exports.IosShellApp = IosShellApp;

const IosCodeSigning = require('./detach/IosCodeSigning.js');

exports.IosCodeSigning = IosCodeSigning;
