// @expo/fastlane/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const turtle_spawn_1 = __importDefault(require("@expo/turtle-spawn"));
async function runFastlane(fastlaneArgs, { logger, envs, cwd } = {}) {
    const fastlaneEnvVars = {
        FASTLANE_DISABLE_COLORS: '1',
        FASTLANE_SKIP_UPDATE_CHECK: '1',
        CI: '1',
        LC_ALL: 'en_US.UTF-8',
        ...envs,
    };
    return await turtle_spawn_1.default('fastlane', fastlaneArgs, {
        env: { ...process.env, ...fastlaneEnvVars },
        logger,
        cwd,
    });
}
async function runFastlaneSafely(fastlaneArgs, logger) {
    try {
        const { stdout, stderr } = await runFastlane(fastlaneArgs, { logger });
        return { stdout, stderr };
    }
    catch (err) {
        const { stdout, stderr } = err;
        return { stdout, stderr, err };
    }
}
exports.runFastlaneSafely = runFastlaneSafely;
exports.default = runFastlane;
